### PR TITLE
Add support for fine-grained access control

### LINKED ISSUE #507: Enable fine grain access per user per workflow or set of workflows
### Is your feature request related to a problem? Please describe

Coming from https://github.com/opensearch-project/opensearch-build/issues/5171
The current jenkins infrastructure has permissions set up which allows only admin or readonly access. See https://github.com/opensearch-project/opensearch-ci/blob/main/lib/compute/oidc-config.ts#L82-L85

### Describe the solution you'd like

Have more fine grain access such that:
- User A can access:
    - Workflow A
    - Workflow B
    - Workflow C
 - User B can access:
     - Workflow A
     - Everything under folder called `foo`
 
- Default read-only and admin access should be retained as is. 
- Everything including user handle, workflow names or pattern or folder and exact permission should be infrastructure as code.

### Describe alternatives you've considered

_No response_

### Additional context

[Role-based Authorization Strategy](https://plugins.jenkins.io/role-strategy/) plugin and [config](https://github.com/opensearch-project/opensearch-ci/blob/main/lib/compute/oidc-config.ts#L79) is already being used. Enhance on top of it to give fine grain access.

### PR DESCRIPTION
**Pull request by gaiksaya**
### Description
This change adds support for fine grained access control. Access can now be controlled at workflow levels per user.
Also fixes the test cases where earlier an already committed Yaml was being parsed. However, we need to actually parse the YAML object. 
Deletes the unused test data.

### Issues Resolved
#507 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: lib/compute/auth-config.ts [hunk 1]
```diff
@@ -103,14 +125,20 @@ export class AuthConfig {
...
           },
-
           ],
         },
       },
     };
 
     jenkinsYaml.jenkins.authorizationStrategy = rolesAndPermissions;
-
+    if (typeof fineGrainedAccessItems !== 'undefined') {
```
Comment (peterzhuamazon) [inline lib/compute/auth-config.ts:134]: Does `!== undefined` better than `=== true` or similar?

File: lib/compute/jenkins-main-node.ts [hunk 2]
```diff
@@ -452,7 +453,8 @@ export class JenkinsMainNode {
     agentNodeObject: AgentNodeConfig, props: AgentNodeNetworkProps, agentNode: AgentNodeProps[], macAgent: string): string {
     let updatedConfig = agentNodeObject.addAgentConfigToJenkinsYaml(stack, agentNode, props, macAgent);
     if (loginAuthProps.authType !== 'default') {
-      updatedConfig = AuthConfig.addOidcConfigToJenkinsYaml(updatedConfig, loginAuthProps.authType, loginAuthProps.adminUsers);
+      updatedConfig = AuthConfig.addOidcConfigToJenkinsYaml(updatedConfig, loginAuthProps.authType,
+        loginAuthProps.adminUsers, loginAuthProps.fineGrainedAccessSpecs);
```
Comment (peterzhuamazon) [inline lib/compute/jenkins-main-node.ts:457]: Will admins have all the permissions defined in above?
And what about read only users?

File: lib/compute/auth-config.ts (hunk 1 above)
Comment (gaiksaya) [reply lib/compute/auth-config.ts:134]: Both have different purpose. One checks if value is undefined and other one checks if value is truly assigned to the variable.

File: lib/compute/jenkins-main-node.ts (hunk 2 above)
Comment (gaiksaya) [reply lib/compute/jenkins-main-node.ts:457]: Admin users will have admin permissions. Same applies for readonly users. The roles are divided into 2 types. global and items at jenkins level. Admin and readonly falls under `global` and fineGrainedAccess falls under `items` category.

**Comment by rishabh6788** [comment]
Can you please explain if I want to add a new user and give them build acess to a particular job, how would I achieve that?

**Comment by gaiksaya** [comment]
> Can you please explain if I want to add a new user and give them build acess to a particular job, how would I achieve that?

Like this https://github.com/opensearch-project/opensearch-ci/pull/514/files#diff-26e449f9955e460d5aabdf24a751a58d9aaace1a73226f143618601a18d0af32R52-R57
The regex pattern defines the access to the workflows. For example:
`((distribution|integ).*)` will give access to all jobs that start with `distrbution-*` and `integ-*`

**Review by peterzhuamazon — APPROVED** [review]

**Comment by peterzhuamazon** [comment]
Asking @rishabh6788 to do another pass before merging.

Thanks!

**Review by rishabh6788 — APPROVED** [review]

**Pull request merged**